### PR TITLE
token bucket: several fixes

### PR DESCRIPTION
**Pull request by mattklein123**
1) Add partial consumption
2) Fix ceiling math for next wakeup time
3) Minor cleanups

Needed for https://github.com/envoyproxy/envoy/issues/5942

Risk Level: Low
Testing: New UT
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by mattklein123** [comment]
@ramaraochavali PTAL

File: source/common/common/token_bucket_impl.cc [hunk 1]
```diff
@@ -17,21 +17,25 @@ bool TokenBucketImpl::consume(uint64_t tokens) {
     last_fill_ = time_now;
   }
 
+  if (allow_partial) {
+    tokens = std::min(tokens, static_cast<uint64_t>(std::floor(tokens_)));
+  }
+
   if (tokens_ < tokens) {
     return false;
```
Comment (ramaraochavali) [inline source/common/common/token_bucket_impl.cc]: nit: should we return zero here to be consistent with return value?

File: include/envoy/common/token_bucket.h [hunk 2]
```diff
@@ -18,16 +18,19 @@ class TokenBucket {
   virtual ~TokenBucket() {}
 
   /**
-   * @param tokens supplies the number of tokens to be consumed. Default is 1.
-   * @return true if bucket is not empty, otherwise it returns false.
+   * @param tokens supplies the number of tokens to be consumed.
+   * @param allow_partial supplies whether the token bucket will allow consumption of less tokens
+   *                      than asked for. If allow_partial is true, the bucket contains 5 tokens,
+   *                      and the caller asks for 3, the bucket will return 3 tokens.
```
Comment (ramaraochavali) [inline include/envoy/common/token_bucket.h]: Should we clarify here what happens to other two tokens?

Comment (mattklein123) [reply include/envoy/common/token_bucket.h]: Oops sorry I typed this wrong, will fix.

File: source/common/common/token_bucket_impl.cc (hunk 1 above)
Comment (mattklein123) [reply source/common/common/token_bucket_impl.cc]: Yup will fix.

File: include/envoy/common/token_bucket.h (hunk 2 above)
Comment (ramaraochavali) [reply include/envoy/common/token_bucket.h]: Thanks. Now it makes sense.

**Pull request merged**